### DISCUSSION TITLE
a native token protocol can be implemented on dogecoin

### DISCUSSION THREAD

**Discussion by dogbitcoin** [dogecoin/dogecoin#3326] [General]
A UTXO-based blockchain structure with a native token method, which I call "dye coins."
1. The transaction output class needs to add a new field called "color".
2. The block header needs to add a new field called "color".
3. The validation of ordinary transactions needs to add additional validation logic.
4. The validation of Coinbase transactions needs to add additional validation logic.
5. The validation of blocks needs to add new validation logic.

Specific logic is as follows:
1. After the transaction output has the color tag field, there are only two results, 0 (all 0) indicates not marked, non-0 indicates marked.
2. To verify that a normal transaction (non-coinbase) is valid, additional validation logic needs to be added. Specifically, it means that the input and output of all transactions must be traversed to ensure that the output values of all colors are less than or equal to the input values of the corresponding colors.
3. Miners need to verify that all or some of the normal transactions in the transaction pool that need to be packaged meet 2, and calculate the difference between all the corresponding colors in all these transactions as utxo (can be single or multiple) added to the Coinbase output.
4. A coinbase transaction is a special transaction that essentially has no input and only output. In addition to the utxo generated by the block miner fee, the coinbase transaction itself has a certain number of native token rewards.
5. The structure of a coinbase transaction that does not participate in color marking is as follows:
* 	Assuming that the current block reward is 50 white coins, and the transaction output generated by the miner fee contains 10 red coins, 10 green coins, and 10 white coins.
* 	When constructing a coinbase transaction, 3 red coin outputs (3 red coin outputs total 10) can be output, 2 green coin outputs (2 green coin outputs total 10) and 1 white coin output (output quantity is equal to 60)
6. The structure of a coinbase transaction that participates in color marking is as follows:
* 	Assuming that the current block reward is 50 white coins, and the transaction output generated by the miner fee contains 10 red coins, 10 green coins, and 10 white coins.
* 	When constructing a coinbase transaction, 3 red coin outputs (3 red coin outputs total 10) can be output, 2 green coin outputs (2 green coin outputs total 10), and the remaining 60 white coins will be dyed yellow as a single output (or multiple outputs). However, 60 white coins cannot be dyed yellow, 30 blue, and white coins can only be dyed the same color and a color that has not appeared before (how to determine if the color exists will be mentioned later), otherwise the transaction will fail.
7. Next, the miner will pack all the transactions to generate the Merkle tree, and assign the color value of the dye in the coinbase transaction to the color field in the block header, and then start Proof of Work to generate the block.
8. After the miner generates the block, it will broadcast the block to the entire network.
9. Other miners will verify the validity of the block. This includes verifying the validity of each transaction, especially the coinbase transaction. It must be ensured that the only color that can be dyed in the coinbase transaction is one color (not that there is only one color of output), and that this color has never appeared before (it can be obtained by traversing all the block headers or caching all the colors locally). Otherwise, the transaction will fail. If there is no dyeing (meaning that the output of the coinbase transaction does not produce a new color utxo, and the number of utxo tokens of other colors does not increase), the traversal process can be skipped.
10. Verify whether the color value in the block header (0 if no dyeing) is equal to the new color value in the coinbase transaction. If equal, the block is valid, the block is added to the blockchain, and the next block is mined. If not equal, the block is invalid and the block is rejected.
11. [Note] There is no inflation problem here, because inflation can be achieved simply by adding one more color utxo, for example, green and red are marked as US dollars.

Implementation: The block header and transaction data structures themselves have a built-in version number field. Version number increases can be used to implement upgrades, which are also part of the protocol.

**Comment by patricklodder** [comment]
What would be the benefit?

**Reply by dogbitcoin** [reply]
Advantages:

Legitimacy is guaranteed by the behavior of node software, which is the responsibility of the system, not the script.
This means that the validation of colored coins is no longer dependent on scripts, but is done by node software. This makes the validation of colored coins more efficient and reliable.
Supports massive concurrency, the independence of utxo ensures that such a token scheme supports high concurrency.
This means that colored coins can support a large number of transactions at the same time without causing performance bottlenecks.
Miners can be natively identified.
This means that miners can identify their own income through colored coins.
The original chain coin shrinks, and the original chain coin shrinks once for every additional colored coin. Theoretically, the original chain coin can be exhausted, but the block time limits the time it may take to exhaust.
This means that colored coins can achieve deflation, thereby suppressing inflation.
The right to mint (dyeing) is only owned by miners, which increases miner incentives.
This means that the right to mint colored coins is monopolized by miners, thereby increasing miner incentives.
Natively supports the point-to-point cash system (wallets can not only have 5 yuan of RMB, but also 1 dollar of USD, 100 yen).
This means that colored coins can support a variety of assets, thereby meeting the diversified needs of users.
Is equivalent to a one-click chain (imagine that the original chain coins are all exhausted, then all the utxo in the blockchain are colored, and the subset of red transactions can be constructed into a digital signature chain with only red coins, of course, it can also be mixed with other colored coins to complete the payment).
This means that colored coins can be used to build new blockchains, thereby achieving rapid and low-cost blockchain creation.

disadvantage:
It is a major change to the current utxo-based node software.

**Reply by dogbitcoin** [reply]
It will enable dogecoin to become the programable money. once recovered all satoshi's original op_code, we will get a utxo-based scaliable blockchain, no tps problem, riched token ecosystem.

**Reply by daank-c** [reply]
using your model, if I had 30 blue dogecoins, and wanted to send them to you,  would I pay transaction fees in blue dogecoins, or regular dogecoins?

**Reply by dogbitcoin** [reply]
It depends on you, miner can  recognize every color. at first, miner can choose regular dogecoin as miner fee, because blue dogecoins don't have a fixed price, once it has a price(such as blue dogecoins are usdt coin), miner would like to directly accept the usdt. it's not a technic problem, depends on the development of the tokens system. in technique layer, miner can natively recognise blue dogecoins is enough.

**Reply by daank-c** [reply]
would blue dogecoins and regular dogecoins have separate fee calculations?

**Reply by dogbitcoin** [reply]
once miner can recognise different color, they can use the separate fee calculations.

**Reply by daank-c** [reply]
Some things to consider:

(1) what happens if a miner doesn't want to accept blue dogecoins as fee, and a transaction in a proposed block is exchanging blue dogecoins? Does the miner not earn a fee for those transactions? Are those transactions discarded?

(2) If blue dogecoins had a perceived value that is less, or more than regular dogecoins, could miners use this to subvert other miners, or users?

(3) Would censorship of blue dogecoins be possible? 

(4) How much would this implementation obscure fee data being used today on platforms around the world, which are currently being valued in DOGE, rather than bluDOGE, redDOGE, etc.

**Reply by dogbitcoin** [reply]
1. if all the miners don't accept blue coins, we can also use normal dogecoin pay miner fee, but my method natively support blue coins. because blue coin also dogecoin, just like a hydrogen with 0, or 1, 2 neutron, they has the same feature. the total number of blue coin is far less then the normal dogecoin. it's a economic problem, not a technic problem. 
2. less is impossible at first, because blue dogecoin must use normal dogecoin as a base to mint
3. only miner has the ability to mint the bluecoin, so if you can censorship miner, you can censorship blue coin. 
4. correct, but I think fee will calculate use usd, and a miner  supported blue coin is more competitive than a miner without support
My method just a technique solution, can only solve the technic problem, economic problem need to use economic way to solve, such as competitiveness and market.
